src/ResourceManager/Profile/Commands.Profile/Context/SetAzureRMContext.cs

```diff
@@ -58,14 +56,40 @@ protected override void ProcessRecord()
                 AzureRmProfileProvider.Instance.Profile.SetContextWithCache(new AzureContext(Context.Subscription, Context.Account,
                     Context.Environment, Context.Tenant));
             }
-            else
+            else if (ParameterSetName == SubscriptionNameParameterSet)
             {
                 var profileClient = new RMProfileClient(AzureRmProfileProvider.Instance.Profile);
-                profileClient.SetCurrentContext(SubscriptionId, TenantId);
-                if (!string.IsNullOrWhiteSpace(SubscriptionName))
+                AzureSubscription subscription = null;
+                string tenantId = AzureRmProfileProvider.Instance.Profile.Context.Tenant.Id.ToString();
+
+                if (string.IsNullOrWhiteSpace(TenantId))
+                {
+                    WriteWarning(
+                    string.Format(
+                        Resources.CurrentTenantInUse,
+                        tenantId));
+                }
+                else
                 {
-                    AzureRmProfileProvider.Instance.Profile.Context.Subscription.Name = SubscriptionName;
+                    tenantId = TenantId;
                 }
+
+                if (!profileClient.TryGetSubscriptionByName(
+                    tenantId,
+                    SubscriptionName,
+                    out subscription))
+                {
+                    throw new ItemNotFoundException(
+                        string.Format(Resources.SubscriptionNameNotFoundError, SubscriptionName));
+                }
+
+                profileClient.SetCurrentContext(subscription.Id.ToString(), tenantId, verifySubscription: false);
+                AzureRmProfileProvider.Instance.Profile.Context.Subscription.Name = SubscriptionName;
+            }
+            else if (ParameterSetName == SubscriptionIdParameterSet)
+            {
+                var profileClient = new RMProfileClient(AzureRmProfileProvider.Instance.Profile);
```

Comment:
In this case, I think we need to do the same validation - make a call to get the subscription by Id and fill in the name - right now this will create a context with no subscription name.